.circleci/template.yml

```diff
@@ -0,0 +1,978 @@
+    docker:
+      - image: *OTP23
+      - *redis_container
+  otp_24_redis:
+    working_directory: ~/app
+    docker:
+      - image: *OTP24
+      - *redis_container
+  otp_24_redis_minio_rmq:
+    working_directory: ~/app
+    docker:
+      - image: *OTP24
+      - *redis_container
+      - *minio_container
+      - *rabbit_container
+  otp_24_mysql_redis:
+    working_directory: ~/app
+    docker:
+      - image: *OTP24
+      - *redis_container
+      - *mysql_container
+      - *rabbit_container
+  otp_23_pgsql_redis:
+    working_directory: ~/app
+    docker:
+      - image: *OTP23
+      - *redis_container
+      - *pgsql_container
+  otp_24_pgsql_redis:
+    working_directory: ~/app
+    docker:
+      - image: *OTP24
+      - *redis_container
+      - *pgsql_container
+  otp_23_ldap_redis:
+    working_directory: ~/app
+    docker:
+      - image: *OTP23
+      - *redis_container
+      - *ldap_container
+  otp_24_ldap_redis:
+    working_directory: ~/app
+    docker:
+      - image: *OTP24
+      - *redis_container
+      - *ldap_container
+  otp_24_mssql_redis:
+    working_directory: ~/app
+    docker:
+      - image: *OTP24
+      - *redis_container
+      - *mssql_container
+commands:
+  install_erlang:
+    parameters:
+      otp_package:
+        type: string
+        description: Version of the Erlang package to install
+    steps:
+    - run:
+        name: Install Erlang
+        command: |
+          sudo killall -9 apt-get || true && \
+          sudo apt-get install -y esl-erlang=1:<<parameters.otp_package>>
+  fetch_packages:
+    steps:
+    - run:
+        name: Install basic packages
+        command: |
+          sudo killall -9 apt-get || true && \
+          echo "Acquire::ForceIPv4 'true';" | sudo tee -a /etc/apt/apt.conf.d/99force-ipv4 && \
+          wget https://packages.erlang-solutions.com/erlang-solutions_2.0_all.deb && \
+          sudo dpkg -i erlang-solutions_2.0_all.deb && \
+          sudo apt-get update && \
+          sudo apt-get install unixodbc-dev -y && \
+          sudo apt-get install unixodbc -y && \
+          sudo apt-get install tdsodbc -y
+  fetch_build_packages:
+    steps:
+      - run:
+          name: Install packages necessary for building
+          command: |
+            sudo killall -9 apt-get || true && \
+            echo "Acquire::ForceIPv4 'true';" | sudo tee -a /etc/apt/apt.conf.d/99force-ipv4 && \
+            sudo apt-get update && \
+            sudo apt-get install libssl-dev unixodbc-dev unixodbc tdsodbc rsync zlib1g-dev -y
+  maybe_build_deps_and_cache:
+    steps:
+      - restore_cache:
+          name: Maybe restore all rebar3 dependencies
+          key: *DEPS_CACHE_KEY
+      - run:
+          name: Get deps
+          command: |
+            tools/configure with-all
+            ./rebar3 get-deps
+      - save_cache:
+          name: Cache downloaded and built dependencies
+          key: *DEPS_CACHE_KEY
+          paths:
+            - ~/.cache/rebar3
```

Comment:
Maybe in this step we also want to cache `_build/default/`, at this point only deps have been compiled, not mongooseim. Also I've realised that deps are compiled it get-deps, and then again compiled when we do rebar3 compile, which is duplicating work.